### ISSUE TITLE
Add a regex recursive strategy narrowing content at each recursion

### ISSUE THREAD

**Issue by scrocquesel** [renovatebot/renovate#12300]
### What would you like Renovate to be able to do?

Recursive should allow to narrow the content at each step. Instead of applying the next regex on the complete match of the previous one, it should be possible to capture an inner content to work on.

This would allow this use case:

```
{
      matchStrings: [
        '"(?<first>[^"]*)":\\s*(?<innerContent>{[^}]*})',
        '"(?<second>[^"]*)":\\s*\\{[^}]*}',
        '"name":\\s*"(?<depName>.*)"[^"]*"type":\\s*"(?<datasource>.*)"[^"]*"value":\\s*"(?<currentValue>.*)"',
      ],
      matchStringsStrategy: 'recursive',
      depNameTemplate: '{{{ first }}}/{{{ second }}}/{{{ depName }}}',
    };
```

on 

```
{
  "backup": {
    "test": {
      "name": "grafana/loki",
      "type": "docker",
      "value": "1.6.1"
    }
  },
  "setup": {
    "test": {
      "name": "python",
      "type": "docker",
      "value": "3.9.0"
    }
  }
}
```

to return depName :

```
backup/test/grafana/loki
setup/test/python
```

instead of

```
backup/backup/grafana/loki
setup/setup/python
```

It may exists a way to capture `second` with a more complicated regex, but it can makes the reading more complicated as the nested level augment.

### If you have any ideas on how this should be implemented, please tell us here.

If a match group with given name (say `innerContent`) exists, `handleRecursive` should recurse on its value instead of `match[0]`.

### Is this a feature you are interested in implementing yourself?

No

**Comment by scrocquesel** [comment]
I'm wondering if this could lead to unwanted replacement.
The captured replace string is the final match only and we will have no clue of the path it takes.

If we take this exemple:
```
{
  "backup": {
    "test": {
      "name": "python",
      "type": "docker",
      "value": "3.9.0"
    }
  },
  "setup": {
    "test": {
      "name": "python",
      "type": "docker",
      "value": "3.9.0"
    }
  }
}
```

We will be able to extract

```
backup/test/python:3.9.0
setup/test/python:3.9.0
```

But during replacement, if backup/test/python:3.90 has a new version, I guess, renovate will replace both occurence of the capture replace string.

````
      "name": "python",
      "type": "docker",
      "value": "3.9.0"
````

which can then lead to not being able to replace a new version of setup/test/python:3.9.0, because the replace string for this match will not exist anymore.

To work around this, I think that replacement strategy should be able to iterate over an array of matched string, and handleRecursive should stack each matched string.

This should also affect the existing code but it is mitigated because depName can be templated only with valid field group name which doesn't give much possibility.

**Comment by viceice** [comment]
Isn't this fixed by #7687 ?

**Comment by scrocquesel** [comment]
> Isn't this fixed by #7687 ?

I have change of project since this requirement, but as far as I remember, we did find a working solution with the mentioned PRs.
